### PR TITLE
feat:add registerFilterHandler

### PR DESCRIPTION
**Pull request by imaxun**
### Checklist
- [x] Code compiles correctly
- [ ] Pull Request has submit to 2020.0 and Greenwich
- [ ] Create at least one junit test if possible
- [x] All tests passing
- [ ] Extend documentation if necessary
- [ ] Add myself / the copyright holder to the AUTHORS file

### CODE REVIEW

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 1]
```diff
@@ -64,18 +68,30 @@ public class PolarisServiceRegistry implements ServiceRegistry<Registration> {
...
-			PolarisDiscoveryHandler polarisDiscoveryHandler,
-			MetadataLocalProperties metadataLocalProperties) {
+								  PolarisDiscoveryHandler polarisDiscoveryHandler,
+								  MetadataLocalProperties metadataLocalProperties) {
 		this.polarisDiscoveryProperties = polarisDiscoveryProperties;
 		this.polarisDiscoveryHandler = polarisDiscoveryHandler;
 		this.metadataLocalProperties = metadataLocalProperties;
 
+		this.registerFilterHandler = new LinkedList<>();
+		this.addRegisterFilterHandler(new DefaultRegisterFilterHandler());
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:89]: How can customized filter injected into registration?

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 2]
```diff
@@ -115,11 +140,16 @@ public void register(Registration registration) {
...
 				heartbeat(heartbeatRequest);
 			}
-		}
-		catch (Exception e) {
+		} catch (Exception e) {
 			log.error("polaris registry, {} register failed...{},",
 					registration.getServiceId(), registration, e);
 			rethrowRuntimeException(e);
+		} finally {
+			registerFilterHandlerInstances.forEach(registerFilterHandler -> {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:148]: Why need to operate the registration request at the end?

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:148]: Why need to operate the registration request at the end?

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 3]
```diff
@@ -143,12 +173,10 @@ public void deregister(Registration registration) {
 		try {
 			ProviderAPI providerClient = polarisDiscoveryHandler.getProviderAPI();
 			providerClient.deRegister(deRegisterRequest);
-		}
-		catch (Exception e) {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:147]: Code style is not correct.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 4]
```diff
@@ -143,12 +173,10 @@ public void deregister(Registration registration) {
...
+		} catch (Exception e) {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:176]: Code style is not correct.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:176]: Code style is not correct.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 5]
```diff
@@ -143,12 +173,10 @@ public void deregister(Registration registration) {
...
 			log.error("ERR_POLARIS_DEREGISTER, de-register failed...{},", registration,
 					e);
-		}
-		finally {
+		} finally {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:179]: Code style is not correct.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:179]: Code style is not correct.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 6]
```diff
@@ -101,7 +117,16 @@ public void register(Registration registration) {
 		instanceRegisterRequest.setMetadata(metadataLocalProperties.getContent());
 		instanceRegisterRequest.setProtocol(polarisDiscoveryProperties.getProtocol());
 		instanceRegisterRequest.setVersion(polarisDiscoveryProperties.getVersion());
+
+		//registerFilterHandlerInstances
+		List<RegisterFilterHandler> registerFilterHandlerInstances = registerFilterHandlerInstances();
 		try {
+			registerFilterHandlerInstances.forEach(registerFilterHandler -> {
+				if (!registerFilterHandler.beforeInvoke(new RegisterFilterHandlerContext(instanceRegisterRequest))) {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:125]: Why not operating instanceRegisterRequest directly but operating RegisterFilterHandlerContext?

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:125]: Why not operating instanceRegisterRequest directly but operating RegisterFilterHandlerContext?

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 7]
```diff
@@ -216,14 +245,32 @@ public void heartbeat(InstanceHeartbeatRequest heartbeatRequest) {
 				}
 
 				polarisDiscoveryHandler.getProviderAPI().heartbeat(heartbeatRequest);
-			}
-			catch (PolarisException e) {
+			} catch (PolarisException e) {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:248]: Code style is not correct.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:248]: Code style is not correct.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 8]
```diff
@@ -216,14 +245,32 @@ public void heartbeat(InstanceHeartbeatRequest heartbeatRequest) {
...
 				log.error("polaris heartbeat[{}]", e.getCode(), e);
-			}
-			catch (Exception e) {
+			} catch (Exception e) {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:250]: Code style is not correct.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:250]: Code style is not correct.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java [hunk 9]
```diff
@@ -216,14 +245,32 @@ public void heartbeat(InstanceHeartbeatRequest heartbeatRequest) {
...
 				log.error("polaris heartbeat runtime error", e);
 			}
 		}, ttl, ttl, TimeUnit.SECONDS);
 	}
 
+	private List<RegisterFilterHandler> registerFilterHandlerInstances() {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:256]: Maybe bean injection is more useful.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:256]: Maybe bean injection is more useful.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistryAutoConfiguration.java [hunk 10]
```diff
@@ -45,30 +52,44 @@
...
 		AutoServiceRegistrationAutoConfiguration.class,
-		PolarisDiscoveryAutoConfiguration.class })
+		PolarisDiscoveryAutoConfiguration.class})
 public class PolarisServiceRegistryAutoConfiguration {
 
 	@Bean
-	public PolarisServiceRegistry polarisServiceRegistry(
+	@ConditionalOnMissingBean
+	protected List<RegisterFilterHandler> registerFilterHandler() {
+		return Lists.newLinkedList();
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistryAutoConfiguration.java:63]: What does this bean use for?

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistryAutoConfiguration.java:63]: What does this bean use for?

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/DefaultRegisterFilterHandler.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultRegisterFilterHandler implements RegisterFilterHandler {
+	private static final Logger log = LoggerFactory
+			.getLogger(DefaultRegisterFilterHandler.class);
+
+	@Override
+	public boolean beforeInvoke(RegisterFilterHandlerContext registerFilterHandlerContext) {
+		log.info("defaultRegisterFilterHandler beforeInvoke info");
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/DefaultRegisterFilterHandler.java:12]: Maybe change log level to debug, or just delete DefaultRegisterFilterHandler. I think this is just used for test for the time being.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/DefaultRegisterFilterHandler.java:12]: Maybe change log level to debug, or just delete DefaultRegisterFilterHandler. I think this is just used for test for the time being.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandler.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package com.tencent.cloud.polaris.registry.filter;
+
+
+public interface RegisterFilterHandler {
+
+	boolean beforeInvoke(RegisterFilterHandlerContext registerFilterHandlerContext);
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandler.java:6]: Add comment for explaining what do these method do.

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandler.java:6]: Add comment for explaining what do these method do.

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java [hunk 13]
```diff
@@ -0,0 +1,15 @@
+package com.tencent.cloud.polaris.registry.filter;
+
+import com.tencent.polaris.api.rpc.InstanceRegisterRequest;
+
+public class RegisterFilterHandlerContext {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java:5]: Why use wrapper?

Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java:5]: Why use wrapper?

**Review by SkyeBeFreeman — CHANGES_REQUESTED** [review]

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/DefaultRegisterFilterHandler.java (hunk 11 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/DefaultRegisterFilterHandler.java:12]: 可以调整日志为调试

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java (hunk 13 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java:5]: 方便后续扩展

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java (hunk 2 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:148]: 一个是注册之前的处理一个是注册之后的

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java (hunk 1 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:89]: 可以通过spring @bean自动装配PolarisServiceRegistryAutoConfiguration中有加对应的实现demo

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java (hunk 13 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/filter/RegisterFilterHandlerContext.java:5]: 为了后续如果有扩展都是用包装类传递

File: spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java (hunk 4 above)
Comment (imaxun) [reply spring-cloud-starter-tencent-polaris-discovery/src/main/java/com/tencent/cloud/polaris/registry/PolarisServiceRegistry.java:176]: 不理解我用的自动格式化

**Comment by SkyeBeFreeman** [comment]
@imaxun Plz fix the conflict.

**Pull request closed without merging**